modules/pam_unix/unix_chkpwd.c

```diff
@@ -138,8 +138,11 @@ int main(int argc, char *argv[])
 	  /* if the caller specifies the username, verify that user
 	     matches it */
 	  if (user == NULL || strcmp(user, argv[1])) {
-	    /* no match -> permanently change to the real user and proceed */
-	    if (setuid(getuid()) != 0)
+	    uid_t ruid = getuid();
+	    gid_t rgid = getgid();
+
+	    /* no match -> permanently change to the real user and group, check for no-return, and proceed */
+	    if (setgid(rgid) != 0 || (rgid != 0 && setgid(0) != -1) || setuid(ruid) != 0 || (ruid != 0 && setuid(0) != -1))
```

Comment:
Why `setgid(0)` and `setuid(0)`?
If `rgid != 0`, this code would do `setgid(rgid)` followed by `setgid(0)`, what's the purpose of that?
Likewise with `ruid`.  This doesn't look correct.

Reply:
Is a sanity check that privileges are really gone (they are expect to fail (`!= -1`), see https://wiki.sei.cmu.edu/confluence/display/c/POS37-C.+Ensure+that+privilege+relinquishment+is+successful.